pkg/plugin/connector/builtin/source.go

```diff
@@ -0,0 +1,127 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/conduitio/conduit-connector-protocol/pconnector"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/plugin/connector"
+	"github.com/rs/zerolog"
+)
+
+// sourcePluginAdapter implements the source plugin interface used internally in
+// Conduit and relays the calls to a source plugin defined in
+// conduit-connector-protocol (pconnector). This adapter needs to make sure it
+// behaves in the same way as the standalone plugin adapter, which communicates
+// with the plugin through gRPC, so that the caller can use both of them
+// interchangeably.
+type sourcePluginAdapter struct {
+	impl pconnector.SourcePlugin
+	// logger is used as the internal logger of sourcePluginAdapter.
+	logger log.CtxLogger
+	// ctxLogger is attached to the context of each call to the plugin.
+	ctxLogger zerolog.Logger
+}
+
+var _ connector.SourcePlugin = (*sourcePluginAdapter)(nil)
+
+func newSourcePluginAdapter(impl pconnector.SourcePlugin, logger log.CtxLogger) *sourcePluginAdapter {
+	return &sourcePluginAdapter{
+		impl:      impl,
+		logger:    logger.WithComponent("builtin.sourcePluginAdapter"),
+		ctxLogger: logger.WithComponent("plugin").ZerologWithComponent(),
+	}
+}
+
+func (s *sourcePluginAdapter) withLogger(ctx context.Context) context.Context {
+	return s.ctxLogger.WithContext(ctx)
+}
+
+func (s *sourcePluginAdapter) Configure(ctx context.Context, in pconnector.SourceConfigureRequest) (pconnector.SourceConfigureResponse, error) {
+	s.logger.Debug(ctx).Any("request", in).Msg("calling Configure")
+	out, err := runSandbox(s.impl.Configure, s.withLogger(ctx), in.Clone(), s.logger)
+	return out.Clone(), err
```

Comment:
Also all methods that call directly into the builtin plugin clone the request and response.

Reply:
What's the issue what cloning solves? We should probably leave a comment somewhere about that.

Reply:
We wouldn't have to clone _all_ of them, as some are empty. The point is that we don't want to share memory between Conduit and the plugin so changes are isolated. For instance, the plugin could manipulate the config map, but we don't want that to change the map in Conduit (that actually happens, when the plugin sets the defaults). In the end, this ensures the plugin works the same way as standalone or builtin.

Reply:
Added a comment https://github.com/ConduitIO/conduit/pull/1622/commits/b708694acd1ee6d25962b5dd3d0e8228c71153e7 👍